torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -157,16 +157,59 @@ void RRefContext::delUser(
     const worker_id_t owner,
     const RRefId& rrefId,
     const ForkId& forkId) {
-  std::lock_guard<std::mutex> lock(destroyedMutex_);
-  if (!destroyed_) {
-    auto fm = agent_->send(
-        agent_->getWorkerInfo(owner),
-        RRefUserDelete(rrefId, forkId).toMessage());
+  {
+    std::lock_guard<std::mutex> lock(destroyedMutex_);
+    if (!destroyed_) {
+      auto fm = agent_->send(
+          agent_->getWorkerInfo(owner),
+          RRefUserDelete(rrefId, forkId).toMessage());
+
+      fm->addCallback([](const Message& /* unused */,
+                         const c10::optional<utils::FutureError>& futErr) {
+        handleException(futErr);
+      });
+    }
+  }
 
-    fm->addCallback([](const Message& /* unused */,
-                       const c10::optional<utils::FutureError>& futErr) {
-      RRefContext::handleException(futErr);
-    });
+  std::lock_guard<std::mutex> lock(mutex_);
+  confirmedUsers_.erase(forkId);
+}
+
+void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
+  // First, wait for all pending UserRRefs to be confirmed,
+  // one kind is pendingUsers_, which are shared from Owner,
+  // the other kind pendingChildren_, which are shared from another User.
+  std::unique_lock<std::mutex> lock(mutex_);
+  bool noPending = deleteAllUsersCV_.wait_for(lock, timeoutMillis, [this]() {
+    return pendingUsers_.size() == 0 && pendingChildren_.size() == 0;
+  });
+  if (!noPending) {
+    LOG(ERROR)
+        << "Timed out waiting for pending UserRRefs to be confirmed by owner and parent.";
+  }
+
+  // Start sending UserRRef delete messages, after all pendings are confirmed.
+  // Note, there should be no new forkings in between, because it's assumed that
+  // this utility is called during graceful shutdown, where no new user RPCs can
+  // not be initiaited anymore.
+  auto tempConfirmedUsers = confirmedUsers_;
```

Comment:
as erasing non-existing keys from unordered_map is fine, we can swap here?

Reply:
Yeah, we can. It's more performant, like move semantics, but bi-directional. 
Also, actually no one is expected to insert to confirmedUsers_ again, based on the assumption that it's called during graceful shutdown.